File: VB/Busy/ProgressControl.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports System.Threading

Namespace Busy

    Public Partial Class ProgressControl
        Inherits UserControl

        Private _IsOperationInProgress As Boolean

        Public Property IsOperationInProgress As Boolean
            Get
                Return _IsOperationInProgress
            End Get

            Set(ByVal value As Boolean)
                _IsOperationInProgress = value
                simpleButton1.Enabled = Not value
            End Set
        End Property

        Private Const INT_filesCount As Integer = 500

        Private Const INT_operationTime As Integer = 10

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Shared Sub CopyFile()
            Thread.Sleep(INT_operationTime)
        End Sub

        Private Sub UpdateProgressBar(ByVal value As Integer)
            progressBarControl1.EditValue = value
        End Sub

        Private Sub backgroundWorker_DoWork(ByVal sender As Object, ByVal e As DoWorkEventArgs)
            For i As Integer = 1 To INT_filesCount
                Call CopyFile()
                Dim progress As Integer = i * 100 \ INT_filesCount
                backgroundWorker.ReportProgress(progress)
            Next
        End Sub

        Private Sub backgroundWorker_ProgressChanged(ByVal sender As Object, ByVal e As ProgressChangedEventArgs)
            UpdateProgressBar(e.ProgressPercentage)
        End Sub

        Private Sub backgroundWorker_Completed(ByVal sender As Object, ByVal e As RunWorkerCompletedEventArgs)
            UpdateProgressBar(0)
            IsOperationInProgress = False
        End Sub

        Private Sub simpleButton1_Click(ByVal sender As Object, ByVal e As EventArgs)
            IsOperationInProgress = True
            backgroundWorker.RunWorkerAsync()
        End Sub
    End Class
End Namespace

File: CS/Busy/ProgressControl.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Threading;

namespace Busy {
    public partial class ProgressControl : UserControl {
        private bool _IsOperationInProgress;
        public bool IsOperationInProgress {
            get { return _IsOperationInProgress; }
            set { _IsOperationInProgress = value; simpleButton1.Enabled = !value; }
        }

        private const int INT_filesCount = 500;
        private const int INT_operationTime = 10;
        public ProgressControl() {
            InitializeComponent();
        }

        private static void CopyFile() {
            Thread.Sleep(INT_operationTime);
        }

        private void UpdateProgressBar(int value) {
           progressBarControl1.EditValue = value;
        }

        private void backgroundWorker_DoWork(object sender, DoWorkEventArgs e) {
            for(int i = 1; i <= INT_filesCount; i++) {
                CopyFile();
                int progress = i * 100 / INT_filesCount;
                backgroundWorker.ReportProgress(progress);
            }
        }


        private void backgroundWorker_ProgressChanged(object sender, ProgressChangedEventArgs e) {
            UpdateProgressBar(e.ProgressPercentage);
        }

        private void backgroundWorker_Completed(object sender, RunWorkerCompletedEventArgs e) {
            UpdateProgressBar(0);
            IsOperationInProgress = false;
        }

        private void simpleButton1_Click(object sender, EventArgs e) {
            IsOperationInProgress = true;
            backgroundWorker.RunWorkerAsync();
        }
    }
}
